algo-core/scheduler_utils.py

```diff
@@ -0,0 +1,148 @@
+                start = i
+                continue
+
+            # End of range:
+            # (A range will end if either the current slot is unavailable
+            # (value 0) or if the current slot is the last one.)
+            if start is not None:
+                if value == 0:  # Unavailable
+                    day_ranges.append([start, i])
+                    start = None
+                elif i == end_hour - 1:  # Last slot
+                    day_ranges.append([start, end_hour])
+                else:
+                    continue
+
+        week_ranges.append(day_ranges)
+
+    return week_ranges
+
+
+def print_final_schedules(schedule_results, df_agents, num_days):
+    """Print final schedule, validate output JSON, and write to file."""
+    for d in range(num_days):
+        print(
+            f"\n{schedule_results[d]['start_date'].strftime('%Y-%m-%d')} "
+            "shifts:"
+        )
+
+        for (i, e) in enumerate(schedule_results[d]["shifts"]):
+            print(e)
+
+    output_json = []
+
+    for epoch in schedule_results:
+        # Substitute agent info from "handle" to "handle <email>"
+        shifts = []
+
+        for (name, start, end) in epoch["shifts"]:
+            shifts.append(
+                {
+                    "agent": f"{name} <{df_agents.loc[name, 'email']}>",
+                    "start": start,
+                    "end": end,
+                }
+            )
+
+        day_dict = {"start_date": epoch["start_date"].strftime("%Y-%m-%d"), "shifts": shifts}
+        output_json.append(day_dict)
+
+    output_json_schema = json.load(
+        open(Path(get_project_root(), "lib", "schemas", "support-shift-scheduler-output.schema.json")))
+
+    try:
+        jsonschema.validate(output_json, output_json_schema)
+    except jsonschema.exceptions.ValidationError as err:
+        print("Output JSON validation error", err)
+        sys.exit(1)
+
+    print("\nSuccessfully validated JSON output.")
+
+    with open(Path(input_folder, "support-shift-scheduler-output.json"), "w") as outfile:
+        outfile.write(json.dumps(output_json, indent=4))
+
+    return output_json
+
+
+def read_onboarding_files():
+    """Read agent handles from onboarding-related files into pandas series."""
+
+    if (path := Path(input_folder, filename_onboarding)).exists():
+        ser_o = pd.read_csv(path, squeeze=True, header=None, names=["agents"])
+    else:
+        ser_o = pd.Series(data=None, name="agents", dtype="str")
+
+    if (path := Path(input_folder, filename_mentors)).exists():
+        ser_o = pd.read_csv(path, squeeze=True, header=None, names=["agents"])
+    else:
+        ser_m = pd.Series(data=None, name="agents", dtype="str")
+
+    if (path := Path(input_folder, filename_new)).exists():
+        ser_n = pd.read_csv(path, squeeze=True, header=None, names=["agents"])
+    else:
+        ser_n = pd.Series(data=None, name="agents", dtype="str")
+
+    return [ser_o, ser_m, ser_n]
```

Comment:
I think the pattern is becoming pretty clear now and could be easily refactored to use the same logic for all onboarding files.

```suggestion
def read_onboarding_file(path: Path) -> pd.Series:
    """Read agent handles from an onboarding-related file into a pandas series."""
    try:
        series = pd.read_csv(path, squeeze=True, header=None, names=["agents"])
    except FileNotFoundError:
        series = pd.Series(data=None, name="agents", dtype="str")
    return series


def read_onboarding_files():
    """Parse all agent handles."""
    return [read_onboarding_file(Path(input_folder, f)) for f in [filename_onboarding, filename_mentors, filename_new]]
```

Note that I used a try-except block instead of explicitly checking for the existence of a file. That's because the [easier to ask for forgiveness than permission](https://docs.python.org/3/glossary.html#term-eafp) paradigm is normally preferred in python.

Reply:
It can also be left as is, I just provided a suggestion in line with my coding style :)